crates/worker/src/partition_processor_manager.rs

```diff
@@ -684,68 +759,75 @@ impl<T: TransportConnect> PartitionProcessorManager<T> {
-                        partition_key_range,
-                        RunMode::Follower,
-                    )
-                    .await?;
+            ProcessorCommand::Follower | ProcessorCommand::Leader => {
+                let run_mode = if control_processor.command == ProcessorCommand::Leader {
+                    RunMode::Leader
+                } else if control_processor.command == ProcessorCommand::Follower {
+                    RunMode::Follower
                 } else {
-                    debug!("Unknown partition id '{partition_id}'. Ignoring follower command.");
-                }
-            }
-            ProcessorCommand::Leader => {
-                if let Some(state) = self.running_partition_processors.get_mut(&partition_id) {
-                    state
-                        .run_for_leader(
-                            self.metadata_store_client.clone(),
-                            self.metadata.my_node_id(),
-                        )
-                        .await?;
+                    unreachable!()
+                };
+
+                if let Some(status) = self.running_partition_processors.get_mut(&partition_id) {
+                    if let ProcessorStatus::Started(state) = status {
+                        // if we error here, then the system is shutting down
+                        if run_mode == RunMode::Follower {
+                            state.step_down()?;
+                        } else if run_mode == RunMode::Leader {
+                            state
+                                .run_for_leader(
+                                    self.metadata_store_client.clone(),
+                                    self.metadata.my_node_id(),
+                                )
+                                .await?;
+                        }
+                    }
+                    // otherwise if it's still starting we can't control it yet
                 } else if let Some(partition_key_range) = partition_table
                     .get_partition(&partition_id)
                     .map(|partition| &partition.key_range)
                 {
-                    self.start_partition_processor(
+                    let starting = self.start_partition_processor(
                         partition_id,
-                        partition_key_range,
-                        RunMode::Leader,
-                    )
-                    .await?;
+                        partition_key_range.clone(),
+                        run_mode,
+                        events,
+                    );
+                    let task_id = self.task_center.spawn_child(
+                        TaskKind::Disposable,
+                        "starting-partition-processor",
+                        Some(partition_id),
+                        starting.instrument(
+                            debug_span!("starting_partition_processor", partition_id=%partition_id),
+                        ),
+                    )?;
```

Comment:
At least on my machine where `starting` seems to contend on some Tokio lock when creating new runtimes, it is a problem running the `starting` task on a Tokio thread. The problem is if enough PPs are started concurrently, they will block all Tokio threads and therefore lead to starvation of the event loop thread. What seems to work on my machine is to run the `starting` future on a blocking thread (`spawn_blocking`).

Reply:
I think it's okay to use `spawn_blocking` instead. On my machine using spawn/spawn_child does not show any problem (for either startup time or on with tokio-console)